File: core/src/nl/t64/game/rpg/screens/world/GameMapQuestBlocker.java
package nl.t64.game.rpg.screens.world;

import com.badlogic.gdx.maps.MapObject;
import com.badlogic.gdx.maps.objects.RectangleMapObject;
import com.badlogic.gdx.math.Rectangle;
import com.badlogic.gdx.math.Vector2;
import nl.t64.game.rpg.Utils;
import nl.t64.game.rpg.subjects.BlockObserver;

import java.util.Optional;


class GameMapQuestBlocker extends GameMapObject implements BlockObserver {

    private final String questId;
    private final boolean isActiveIfComplete;
    private final String taskId;
    private boolean isActive;

    GameMapQuestBlocker(MapObject mapObject) {
        RectangleMapObject rectObject = (RectangleMapObject) mapObject;

        this.rectangle = rectObject.getRectangle();
        this.questId = rectObject.getProperties().get("type", String.class);
        this.isActiveIfComplete = rectObject.getProperties().get("activeIfComplete", Boolean.class);
        this.taskId = rectObject.getProperties().get("task", String.class);
        this.isActive = rectObject.getProperties().get("isActive", Boolean.class);

        if (this.isActive) {
            Utils.getBrokerManager().blockObservers.addObserver(this);
        }
    }

    @Override
    public Optional<Rectangle> getBlockerFor(Rectangle boundingBox) {
        if (isActive && boundingBox.overlaps(rectangle)) {
            return Optional.of(rectangle);
        } else {
            return Optional.empty();
        }
    }

    @Override
    public boolean isBlocking(Vector2 point) {
        return isActive && rectangle.contains(point);
    }

    void update() {
        final boolean isFinished = Utils.getGameData().getQuests().getQuestById(questId).isFinished();
        final boolean isComplete = Utils.getGameData().getQuests().getQuestById(questId).isTaskComplete(taskId);

        final boolean before = isActive;
        isActive = (isFinished || isComplete) == isActiveIfComplete;
        final boolean after = isActive;
        if (before != after) {
            changeBlocker();
        }
    }

    private void changeBlocker() {
        if (isActive) {
            Utils.getBrokerManager().blockObservers.addObserver(this);
        } else {
            Utils.getBrokerManager().blockObservers.removeObserver(this);
        }
        Utils.getMapManager().setTiledGraph();
    }

}

File: core/src/nl/t64/game/rpg/screens/world/mapobjects/GameMapQuestBlocker.kt
package nl.t64.game.rpg.screens.world.mapobjects

import com.badlogic.gdx.maps.objects.RectangleMapObject
import com.badlogic.gdx.math.Rectangle
import com.badlogic.gdx.math.Vector2
import ktx.tiled.property
import ktx.tiled.propertyOrNull
import nl.t64.game.rpg.Utils.brokerManager
import nl.t64.game.rpg.Utils.gameData
import nl.t64.game.rpg.Utils.mapManager
import nl.t64.game.rpg.components.quest.QuestState
import nl.t64.game.rpg.subjects.BlockObserver


class GameMapQuestBlocker(rectObject: RectangleMapObject) : GameMapObject(rectObject.rectangle), BlockObserver {

    private val questId: String = rectObject.name
    private val isActiveIfComplete: Boolean = rectObject.property("activeIfComplete")
    private val taskId: String? = rectObject.propertyOrNull("task")
    private var isActive: Boolean = rectObject.property<Boolean>("isActive").also {
        if (it) brokerManager.blockObservers.addObserver(this)
    }

    override fun getBlockerFor(boundingBox: Rectangle): Rectangle? {
        return rectangle.takeIf { isActive && boundingBox.overlaps(it) }
    }

    override fun isBlocking(point: Vector2): Boolean {
        return isActive && rectangle.contains(point)
    }

    fun update() {
        val isFinished = gameData.quests.isCurrentStateEqualOrHigherThan(questId, QuestState.FINISHED)
        val isComplete = gameData.quests.isTaskComplete(questId, taskId)

        val before = isActive
        isActive = (isFinished || isComplete) == isActiveIfComplete
        val after = isActive
        if (before != after) {
            changeBlocker()
        }
    }

    private fun changeBlocker() {
        if (isActive) {
            brokerManager.blockObservers.addObserver(this)
        } else {
            brokerManager.blockObservers.removeObserver(this)
        }
        mapManager.setTiledGraph()
    }

}
